api/odilo.py

```diff
@@ -0,0 +1,495 @@
+        """Request an OAuth bearer token that allows us to act on
+        behalf of a specific patron.
+        """
+
+        self.client_key = patron
+        self.client_secret = pin
+        message = self.refresh_creds(credential)
+
+        if 'OK' == message:
+            return credential
+        else:
+            raise PatronAuthorizationFailedException(message)
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """Check out a book on behalf of a patron.
+
+        :param patron: a Patron object for the patron who wants
+        to check out the book.
+
+        :param pin: The patron's alleged password.
+
+        :param licensepool: Identifier of the book to be checked out is 
+        attached to this licensepool.
+
+        :param internal_format: Represents the patron's desired book format.
+
+        :return: a LoanInfo object.
+        """
+        record_id = licensepool.identifier.identifier
+
+        # We have 2 formats, 'EBOOK_STREAMING' could be in our API ('PDF', 'EPUB'):
+        ebook_streaming = False
+        if internal_format == 'EBOOK_STREAMING':
+            ebook_streaming = True
+            internal_format = 'EPUB'  # Try first ebook_streaming as EPUB
+
+        # Data just as 'x-www-form-urlencoded', no JSON
+        payload = dict(patronId=patron, format=internal_format)
+
+        response = self.patron_request(
+            patron, pin, self.CHECKOUT_ENDPOINT.replace("{recordId}", record_id),
+            extra_headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            data=payload)
+
+        if response.content:
+            response_json = response.json()
+            if response.status_code == 404:
+                error_id, error_description = self.raise_exception_on_error(response_json)
+                raise CannotLoan('Error id: %s, description: %s' % (error_id, error_description))
+            else:
+                return self.loan_info_from_odilo_checkout(licensepool.collection, response_json)
+
+        elif response.status_code == 400 and ebook_streaming:
+            internal_format = 'PDF'  # Retry now with ebook_streaming as PDF
+            return self.checkout(patron, pin, licensepool, internal_format)
+
+        # TODO: we need to improve this at the API and use an error code
+        elif response.status_code == 400:
+            raise NoAcceptableFormat('record_id: %s, format: %s' % (record_id, internal_format))
+
+        raise CannotLoan('patron: %s, record_id: %s, format: %s' % (patron, record_id, internal_format))
+
+    def loan_info_from_odilo_checkout(self, collection, checkout):
+        start_date = self.extract_date(checkout, 'startTime')
+        end_date = self.extract_date(checkout, 'endTime')
+
+        return LoanInfo(
+            collection,
+            DataSource.ODILO,
+            Identifier.ODILO_ID,
+            checkout['id'],
+            start_date,
+            end_date,
+            checkout['downloadUrl']
```

Comment:
I just want to confirm that this downloadUrl is the place you go to get (e.g.) the ACSM file. Is this field always present (but possibly null), even in a streaming context?

Reply:
Yes, you are right. This URL is the place where you go to get an ACSM file, the streaming contents, etc...
This field is always present, and it shouldn't be null if all it's ok. It is specially necessary in a streaming context, if not, you couldn't get the actual streaming resource.

I mean, if you call that URL for a checkout with format:
'ACSM' --> you get the .acsm file to be downloaded.
'EBOOK_STREAMING'  --> you get redirected to the online viewer and get the streaming contents themselves of the resource.
Same for formats video, audio, etc...